examples/advanced/quickxplain.py

```diff
@@ -0,0 +1,82 @@
+"""
+    CPMpy implementation of the QuickXplain algorithm by Junker.
+        Junker, Ulrich. "Preferred explanations and relaxations for over-constrained problems." AAAI-2004. 2004.
+        https://cdn.aaai.org/AAAI/2004/AAAI04-027.pdf
+"""
+
+import cpmpy as cp
+from cpmpy.transformations.get_variables import get_variables
+from cpmpy.transformations.normalize import toplevel_list
+
+
+def quickXplain(soft, hard=[], solver="ortools"):
+    """
+        Find a preferred minimal unsatisfiable subset of constraints
+        A partial order is imposed on the constraints using the order of `soft`.
+        Constraints with lower index are preferred over ones with higher index
+
+        :param: soft: soft constraints, list of expressions
+        :param: hard: hard constraints which cannot be relaxed, optional, list of expressions
+        :param: solver: name of a solver, see SolverLookup.solvernames()
+            "z3", "pysat" and "gurobi" are incremental, "ortools" restarts the solver
+    """
+
+    soft = toplevel_list(soft, merge_and=False)
+    assump = cp.boolvar(shape=len(soft))
+    solver = cp.SolverLookup.get(solver, cp.Model(hard))
+    solver += assump.implies(soft)
+
+    assert solver.solve(assumptions=assump) is False, "The model should be UNSAT!"
+
+    dmap = dict(zip(assump, soft))
+    core = recurse_explain(list(assump),[],[], solver=solver)
+    return [dmap[a] for a in core]
+
+
+def recurse_explain(soft, hard, delta, solver):
```

Comment:
Is there a reason why you recurse_explain (resp. do_recursion)  in the tools is embedded and not here ?
Change the function name to be consistent across both files?

Reply:
Not really, I wanted to stay very close to the original pseudocode of the paper in the example, while the tools-version has some subtle optimizations, plus embedding the recursive call ensures users cannot call the helper function themselves by accident.